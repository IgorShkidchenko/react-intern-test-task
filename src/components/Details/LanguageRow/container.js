import React from "react";

import LanguageRowComponent from "./component";

const LanguageRowContainer = (props) => <LanguageRowComponent {...props}/>;

export default LanguageRowContainer
