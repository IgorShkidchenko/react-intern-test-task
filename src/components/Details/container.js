import React from "react";
import PropTypes from "prop-types";

import DetailsComponent from "./component";

const DetailsContainer = props => <DetailsComponent repository={props.location.state.repository} />;

DetailsContainer.propTypes = {
  location: PropTypes.object.isRequired,
};

export default DetailsContainer;
