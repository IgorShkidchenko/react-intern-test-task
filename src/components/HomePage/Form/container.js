import React from "react"
import axios from "axios"

import FormComponent from "./component"

class FormContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
      valid: true
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    this.setState({value: event.target.value});
  }

  handleSubmit(event) {
    event.preventDefault()

    const url = `https://api.github.com/repos/${this.state.value}`;

    let promiseResponse;
    axios.get(url)
      .then(response => {
        promiseResponse = response.data;

        this.setState({valid: true});
        this.props.onAddRepository(promiseResponse)
      })
      .catch(error => {
        console.log(error)
        this.setState({valid: false});
      })
  }

  render() {
    return <FormComponent
      value={this.state.value}
      valid={this.state.valid}
      handleChange={this.handleChange}
      handleSubmit={this.handleSubmit}
    />;
  }
}

export default FormContainer
