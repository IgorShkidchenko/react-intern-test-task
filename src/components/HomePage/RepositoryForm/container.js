import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Formik, Form, Field, ErrorMessage } from "formik";

import { repositoryActions } from "../../../store/repositories";

export class RepositoryForm extends React.Component {
  render() {
    return (
      <Formik
        initialValues={{ repositoryUrl: "" }}
        validate={(values) => {
          const errors = {};
          if (!values.repositoryUrl) {
            errors.repositoryUrl = "Required";
          } else if (
            !/^.+\/{1}.+$/.test(values.repositoryUrl)
          ) {
            errors.repositoryUrl = "Invalid repository address";
          }
          return errors;
        }}
        onSubmit={(values, { setSubmitting, setErrors }) => {
          const payload = {
            repositoryUrl: values.repositoryUrl,
            setSubmitting,
            setErrors,
          };
          this.props.onAddRepository(payload);
        }}
      >
        {({ isSubmitting }) => (
          <Form>
            <Field type="repositoryUrl" name="repositoryUrl" placeholder="Enter repository name e.g. facebook/react" />
            <ErrorMessage name="repositoryUrl" component="p" />
            <button type="submit" disabled={isSubmitting}>
              Submit
            </button>
          </Form>
        )}
      </Formik>
    );
  }
}

RepositoryForm.propTypes = {
  onAddRepository: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  onAddRepository: payload => dispatch(repositoryActions.getRepository(payload)),
});

export default connect(null, mapDispatchToProps)(RepositoryForm);
