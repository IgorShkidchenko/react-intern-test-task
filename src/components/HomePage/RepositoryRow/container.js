import React from "react";
import PropTypes from "prop-types";

import RepotisoryRowComponent from "./component";

class RepotisoryRowContainer extends React.Component {
  constructor(props) {
    super(props);

    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(event) {
    event.preventDefault();
    this.props.onRemoveRepository(this.props.repository.id);
  }

  render() {
    return (<RepotisoryRowComponent
      repository={this.props.repository}
      handleClick={this.handleClick}
    />);
  }
}

RepotisoryRowContainer.propTypes = {
  repository: PropTypes.object.isRequired,
  onRemoveRepository: PropTypes.func.isRequired,
};

export default RepotisoryRowContainer;
