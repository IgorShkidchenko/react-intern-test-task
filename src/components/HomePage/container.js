import React from "react";

import HomePageComponent from "./component";

class HomePageContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      repositories: []
    }

    this.handleAddRepository = this.handleAddRepository.bind(this);
    this.handleRemoveRepository = this.handleRemoveRepository.bind(this);
  }

  handleAddRepository(repository) {
    const oldRepositories = this.state.repositories;
    oldRepositories.push(repository)

    this.setState({
      repositories: oldRepositories
    });
  }

  handleRemoveRepository(id) {
    const filteredRepositories = this.state.repositories.filter(repo => repo.id !== id);

    this.setState({
      repositories: filteredRepositories
    });
  }

  render() {
    return <HomePageComponent
      self={this}
      repositories={this.state.repositories}
      handleAddRepository={this.handleAddRepository}
      handleRemoveRepository={this.handleRemoveRepository}
      />
  }
}

export default HomePageContainer
