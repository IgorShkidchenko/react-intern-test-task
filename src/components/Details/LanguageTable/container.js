import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import { repositoryActions } from "../../../store/repositories";
import LanguageTableComponent from "./component";
import LanguageRow from ".././LanguageRow";

export class LanguageTableContainer extends React.Component {
  componentDidMount() {
    this.props.loadRepositoryLanguages(this.props.repository.languages_url);
  }

  sumOfObjectValues(hash) {
    let sum = 0;

    Object.values(hash).forEach((languageBytes) => {
      sum += languageBytes;
    });

    return sum;
  };

  calculatePercentage(num, total) {
    return ((num * 100) / total).toFixed(2);
  }

  render() {
    const rows = [];
    const { repositoryLanguages } = this.props;

    if (repositoryLanguages !== null) {
      const totalRepositoryLanguagesBytes = this.sumOfObjectValues(repositoryLanguages);

      Object.entries(repositoryLanguages).forEach((language) => {
        const languageName = language[0];
        const languageBytes = language[1];

        rows.push(<LanguageRow
          percentage={this.calculatePercentage(languageBytes, totalRepositoryLanguagesBytes)}
          languageName={languageName}
          key={languageName}
        />);
      });
    }

    return <LanguageTableComponent rows={rows} />;
  }
}

LanguageTableContainer.propTypes = {
  repository: PropTypes.object.isRequired,
  repositoryLanguages: PropTypes.object.isRequired,
  loadRepositoryLanguages: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  repositoryLanguages: state.repo.repositoryLanguages,
});

const mapDispatchToProps = dispatch => ({
  loadRepositoryLanguages: url => dispatch(repositoryActions.loadRepositoryLanguages(url)),
});

export default connect(mapStateToProps, mapDispatchToProps)(LanguageTableContainer);
