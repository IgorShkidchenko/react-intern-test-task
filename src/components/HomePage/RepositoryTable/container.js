import React from "react";

import RepositoryTableComponent from "./component";
import RepositoryRow from ".././RepositoryRow";

const RepositoryTableContainer = (props) => {
  const rows = [];
  const sortedRepos = props.repositories.sort((a, b) => b.stargazers_count - a.stargazers_count);

  sortedRepos.forEach((repository) => {
    rows.push(
      <RepositoryRow
        repository={repository}
        key={repository.id}
        onRemoveRepository={props.onRemoveRepository} />
    );
  });

  return <RepositoryTableComponent rows={rows} />;
}

export default RepositoryTableContainer
