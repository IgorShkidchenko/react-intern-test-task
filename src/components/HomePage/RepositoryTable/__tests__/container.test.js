import React from "react";
import { shallow } from "enzyme";

import { RepositoryTableContainer } from "../container";

describe("<RepositoryTableContainer />", () => {
  it("renders correctly", () => {
    const repositories = [{ id: 1, stargazers_count: 1 }, { id: 2, stargazers_count: 2 }];
    const wrapper = shallow(<RepositoryTableContainer repositories={repositories} />);

    expect(wrapper).toMatchSnapshot();
  });
});
