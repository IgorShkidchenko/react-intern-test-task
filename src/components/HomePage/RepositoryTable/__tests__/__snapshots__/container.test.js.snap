// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepositoryTableContainer /> renders correctly 1`] = `
<RepositoryTableComponent
  rows={
    Array [
      <Memo(Connect(RepotisoryRowContainer))
        repository={
          Object {
            "id": 2,
            "stargazers_count": 2,
          }
        }
      />,
      <Memo(Connect(RepotisoryRowContainer))
        repository={
          Object {
            "id": 1,
            "stargazers_count": 1,
          }
        }
      />,
    ]
  }
/>
`;
